src/vfs/virtualFileSystemWithWatch.ts

```diff
@@ -0,0 +1,714 @@
+                    this.ensureFolder("/", /*ignoreWatch*/ false);
+                }
+                else {
+                    this.createDirectory(base, recursive);
+                }
+            }
+
+            // base folder has to be present
+            const baseFolder = this.fs.get(base) as FsFolder;
+            Debug.assert(isFsFolder(baseFolder));
+
+            Debug.assert(!this.fs.get(folder.path));
+            this.addFileOrFolderInFolder(baseFolder, folder);
+        }
+
+        writeFile(path: string, content: string): void {
+            const file = this.toFsFile({ path, content });
+
+            // base folder has to be present
+            const base = getDirectoryPath(file.path);
+            const folder = Debug.checkDefined(this.getRealFolder(base));
+
+            if (folder.path === base) {
+                if (!this.fs.has(file.path)) {
+                    this.addFileOrFolderInFolder(folder, file);
+                }
+                else {
+                    this.modifyFile(path, content);
+                }
+            }
+            else {
+                this.writeFile(this.realpath(path), content);
+            }
+        }
+
+        write(_message: string) {
+            // Do nothing
+        }
+
+        realpath(s: string): string {
+            const fullPath = this.toNormalizedAbsolutePath(s);
+            const path = this.toPath(fullPath);
+            if (getDirectoryPath(path) === path) {
+                // Root
+                return s;
+            }
+            const dirFullPath = this.realpath(getDirectoryPath(fullPath));
+            const realFullPath = combinePaths(dirFullPath, getBaseFileName(fullPath));
+            const fsEntry = this.fs.get(this.toPath(realFullPath))!;
+            if (isFsSymLink(fsEntry)) {
+                return this.realpath(fsEntry.symLink);
+            }
+
+            // realpath supports non-existent files, so there may not be an fsEntry
+            return fsEntry?.fullPath || realFullPath;
+        }
+
+        readonly exitMessage = "System Exit";
+        exitCode: number | undefined;
+        readonly resolvePath = (s: string) => s;
+        readonly getExecutingFilePath = () => this.executingFilePath;
+        readonly getCurrentDirectory = () => this.currentDirectory;
+        exit(exitCode?: number) {
+            this.exitCode = exitCode;
+            throw new Error(this.exitMessage);
+        }
+
+        getEnvironmentVariable(_name: string) {
+            return "";
+        }
+
+        snap(): ESMap<Path, FSEntry> {
+            const result = new Map<Path, FSEntry>();
+            this.fs.forEach((value, key) => {
+                const cloneValue = clone(value);
+                if (isFsFolder(cloneValue)) {
+                    cloneValue.entries = cloneValue.entries.map(clone) as SortedArray<FSEntry>;
+                }
+                result.set(key, cloneValue);
+            });
+

```

Comment:
Can you make snap and diff as part of the test static functions that take VFS so its not on tsserver.js